### ISSUE TITLE
Ensure test env components always stop

### ISSUE THREAD

**Issue by lbradstreet** [onyx-platform/onyx#686]
PR #685 was a good start in this direction, however there will be situations where a system such as OnyxDevelopmentEnv is partially started, and will not be able to be shutdown.

This is actually quite hard do right. If you catch all the exceptions and allow the system to start up, you will get even more failures down the line, but if you do not catch the exceptions, then you will not have access to some of the sub-components and will not be able to component/stop them again.

All my prior thoughts dealing with this in system/component ends up getting pretty ugly. What if we catch exceptions in all of the subcomponents, and implement a success? protocol for each component, so we know whether to go further or whether to stop the component again?

cc/ @ChrisBlom @MichaelDrogalis

**Comment by MichaelDrogalis** [comment]
I like that suggestion a lot actually. That seems quite reasonable.

**Comment by mariusz-jachimowicz-83** [comment]
I am thinking about 2 phase start like:
- first phase only to get whole components tree available
- second phase when we initialize dynamic staff like go-loops, futures, ....

I am thinking also about some Supervisor components to be able to mimic in some way erlang style of thinking, and be able to restart/stop component/component after failures

So we could have components, something like:
```
Channels
Supervisor-failures  :depends [Channels]
CompA  :depends [Channels]
Supervisor-CompA :depends [Channels CompA]
CompB  :depends [Channels]
CompC  :depends [Channels]
Supervisor-lifecycle :depends [Channels CompA Supervisor-CompA CompB CompC]
```
`Channels` - holds channels. So each componet can send msg to any other component in system via those channels.
`Supervisor-failures`  - each component is able to notify that failure occured via `failures` channel. This supervisor decide (orchestrate) what to do via sending msg to component/components. Some failures might require restarting couple of components.
`Supervisor-CompA` - allows to restart dynamic parts of CompA like futures, threads, go-loops, ....
`Supervisor-lifecycle` - broadcast msg to all components that they can start `second phase`

This is a little scratch of an experiment that I am doing right now. I am curious how far I can go with this. It's not very clearly describing my idea because I must admit that it is actually quite hard do it right.

**Comment by MichaelDrogalis** [comment]
We tried something quite similar in the past for a subcomponent of Onyx -- we were transitioning from a fully isolated resources per virtual peer to multiplexed, shared resources across an entire peer group. I came up with a design that used a supervisor to monitor component failure and do restarts based on message passing. It is indeed very complicated to do right, and we scrapped it because it was almost impossible to test, so asynchrony is a no go to fix this one. I'll look back and try to find the commit where we had this implemented to compare.

Ideally the patch should work on any Sierra Component-based system.

**Comment by MichaelDrogalis** [comment]
Found it: https://github.com/onyx-platform/onyx/blob/61733d978394c6002f9a4e758f0c0038f5b3a0bd/src/onyx/peer/supervisor.clj

**Comment by MichaelDrogalis** [comment]
@mariusz-jachimowicz-83 That all said, I still think it's a good idea to have a general Clojure library that can do Component supervision. I don't think it's a good fit for Onyx, but there are plenty of use cases that could benefit from Component supervision -- for example, one Component that depends on another that occasionally experiences ephemeral failure. Those situations can certainly benefit from the design that you outlined.

**Comment by mariusz-jachimowicz-83** [comment]
@MichaelDrogalis Sure. I am doing some experiments and learn from it because it is simpler than trying to write general library at once. I will look at your supervisor experiment.

**Comment by lbradstreet** [comment]
@mariusz-jachimowicz-83 @MichaelDrogalis I have a branch to try and make things start more safely here. I've made a few other changes like splitting the zookeeper log component which writes the required paths, from the server component.

The key innovation is catching the component exceptions, and pulling the partially started system from the ex-data. We can then shut it down again as seen here: https://github.com/onyx-platform/onyx/commit/2f4da124dfe38713ce016e6407f3d8a885528983#diff-4f447e1c4650c467574de7f5b160da8dR112

This will require changing all of the components to check whether the component was started, e.g. with a :started? key in every component. If it hasn't been :started? then shutdown becomes a no-op.

Unfortunately, the ZK server port binding is still eluding me when I ctrl-C as zookeeper is starting up. I can get a TestingServer object and stop it but unfortunately the port is still bound.

I'm going to back off work on it for a bit. I think it's better, but I would want to test it more thoroughly in dev before bringing it in to master.

**Comment by MichaelDrogalis** [comment]
Thanks, this is what I had in mind. Also thank you for preemptively splitting out the ZK server, since that needed to be done to clear the way for storing job data in S3.

**Comment by mariusz-jachimowicz-83** [comment]
Eh...one problem with current components library is that it creates children before parents. If I want to create supervisors tree than I can't start supervisor before children components.

**Comment by MichaelDrogalis** [comment]
@mariusz-jachimowicz-83 Yep. You were on the right track with needing a 2-phased start in order to use Component or Mount. I had to deal with the same thing -- a "cold start" phase to set up the tree, and then top-down "warm start".

**Comment by lbradstreet** [comment]
@mariusz-jachimowiz-83 as a side note, you may be interested in this article about how erlang supervisors work 

http://jbodah.github.io/blog/2016/11/18/supervisors-work/

**Comment by mariusz-jachimowicz-83** [comment]
very interesting library https://github.com/xsc/peripheral
